domain/secret/service/service.go

```diff
@@ -647,3 +658,38 @@ func (s *SecretService) ChangeSecretBackend(ctx context.Context, uri *secrets.UR
 	// TODO(secrets)
 	return nil
 }
+
+// SecretRotated rotates the secret with the specified URI.
+func (s *SecretService) SecretRotated(ctx context.Context, uri *secrets.URI, params SecretRotatedParams) error {
+	if err := s.canManage(ctx, uri, params.Accessor, params.LeaderToken); err != nil {
+		return errors.Trace(err)
+	}
+
+	md, err := s.GetSecret(ctx, uri)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	if !md.RotatePolicy.WillRotate() {
+		s.logger.Debugf("secret %q was rotated but now is set to not rotate")
+		return nil
+	}
+	lastRotateTime := md.NextRotateTime
+	if lastRotateTime == nil {
+		now := s.clock.Now()
+		lastRotateTime = &now
+	}
+	nextRotateTime := *md.RotatePolicy.NextRotateTime(*lastRotateTime)
+	s.logger.Debugf("secret %q was rotated: rev was %d, now %d", uri.ID, params.OriginalRevision, md.LatestRevision)
+	// If the secret will expire before it is due to be next rotated, rotate sooner to allow
+	// the charm a chance to update it before it expires.
+	willExpire := md.LatestExpireTime != nil && md.LatestExpireTime.Before(nextRotateTime)
+	forcedRotateTime := lastRotateTime.Add(secrets.RotateRetryDelay)
+	if willExpire {
+		s.logger.Warningf("secret %q rev %d will expire before next scheduled rotation", uri.ID, md.LatestRevision)
+	}
+	if willExpire && forcedRotateTime.Before(*md.LatestExpireTime) || !params.Skip && md.LatestRevision == params.OriginalRevision {
```

Comment:
I can't recall a situation where LatestRevision will be different to OriginalRevision - the charm caller I think always sets OriginalRevision to LatestRevision so I think it would be good to re-evaluate the need for the `md.LatestRevision == params.OriginalRevision` check as it will allow one less bit of data to be queries in performing this logic.

Reply:
I think the LatestRevision - OriginalRevision check is needed if the charm both updates the secret and rotates it in the same hook - worth checking to confirm. I wonder if we can optimise the logic to avoid the latest revision look up.

Reply:
Yeah, the OriginalRevision is LatestRevision. And the `params.Skip` is never set to true.
Does it mean here we always set the nextRotateTime to forcedRotateTime?

Reply:
It does seem like that's the case. Looking at the spec, the ability to skip a rotation was removed from the spec. But the spec does say "Until the charm updates the secret to create a new revision, juju will continue to regularly call the rotate hook to notify that the secret is supposed to be updated.". So we need to ensure that behaviour is in place.

Reply:
So the revision passed in from the secret cached in the  hook context will be different from the current latest revision if the charm calls secret-set during the rotated hook. So the logic looks correct expect that Skip is not implemented.

Reply:
Right.